### Exercise 3.7, problem 10

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn import linear_model

# Load StatsModels API
import statsmodels.api as sm
import statsmodels.formula.api as smf

df = pd.read_csv('Data/Carseats.csv');
df

Sales  CompPrice  Income  Advertising  Population  Price ShelveLoc  Age  \
0     9.50        138      73           11         276    120       Bad   42   
1    11.22        111      48           16         260     83      Good   65   
2    10.06        113      35           10         269     80    Medium   59   
3     7.40        117     100            4         466     97    Medium   55   
4     4.15        141      64            3         340    128       Bad   38   
..     ...        ...     ...          ...         ...    ...       ...  ...   
395  12.57        138     108           17         203    128      Good   33   
396   6.14        139      23            3          37    120    Medium   55   
397   7.41        162      26           12         368    159    Medium   40   
398   5.94        100      79            7         284     95       Bad   50   
399   9.71        134      37            0          27    120      Good   49   

     Education Urban   US  
0           17   Yes  Yes  
1           10   Yes  Yes  
2           12   Yes  Yes  
3           14   Yes  Yes  
4           13   Yes   No  
..         ...   ...  ...  
395         14   Yes  Yes  
396         11    No  Yes  
397         18   Yes  Yes  
398         12   Yes  Yes  
399         16   Yes  Yes  

[400 rows x 11 columns]

#### PART A : 
Fit a multiple regression model to predict Sales using Price, Urban, and US.

In [2]:
#Load data and convert string to integers for analysis
#df.Urban = df.Urban.map(dict(Yes=1, No=0))
#df.US = df.US.map(dict(Yes=1, No=0))

#Create multiple regression model to predict values of sales utilizing R Style Functions
#Method learned from https://datatofish.com/statsmodels-linear-regression/ 
mod_1 = smf.ols(formula = "Sales ~ Price + Urban + US", data = df)
res_1 = mod_1.fit()
print(res_1.summary())

                            OLS Regression Results                            
Dep. Variable:                  Sales   R-squared:                       0.239
Model:                            OLS   Adj. R-squared:                  0.234
Method:                 Least Squares   F-statistic:                     41.52
Date:                Mon, 19 Sep 2022   Prob (F-statistic):           2.39e-23
Time:                        18:10:59   Log-Likelihood:                -927.66
No. Observations:                 400   AIC:                             1863.
Df Residuals:                     396   BIC:                             1879.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                   coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------
Intercept       13.0435      0.651     20.036   

#### PART B : 
Provide an interpretation of each coefficient in the model. Be careful—some of the variables in the model are qualitative.

The coefficient for Price is *-0.0545*, which represents that at a certain fixed value for Urban and US, thus a specific location, increasing the price of a carset by 1 would result in a decrease in sales of 54.45 on average. </br></br>
The coefficient for Urban - Yes is *-0.0219*, which represents the at a given constant price and US, the urban areas will have approximately 22 less sales than non-urban areas. </br> </br>
The coefficient for US - Yes is 1.200537, which means at a constant price and value for Urban, there are 1201 more units on average inside the US over outside

#### PART C : 
Write out the model in equation form, being careful to handle
the qualitative variables properly.!

The model would be </br>
*Ŷ = Intercept + (Price coeff)X<sub>1</sub> + (Urban Coeff)X<sub>2</sub> + (US Coeff)X<sub>3</sub>* </br>
Where Ŷ is the estimate of carseat sales (in thousands), X1 is the price of a carseat in a store, X2 is a whether a carseat is in an urban area, and X2 is whether a carseat is inside the US. The last two values are produced by returning 0 if they are not located in above specified area, and 1 if they are. </br>
Therefore our estimated model would be : </br>
*Ŷ = 13.0435 - 0.0545X<sub>1</sub> - 0.0219X<sub>2</sub> + 1.200537X<sub>3</sub>*

#### PART D : 
For which of the predictors can you reject the null hypothesis H0 : βj = 0?

In order to reject the null hypothesis, we need to have low p-values for all coefficients. The p-values we found were : </br>
**Intercept :** 0.000</br>
**Urban - Yes :** 0.936</br>
**US - Yes :** 0.000</br>
**Price :** 0.000</br>
Because Intercept, US, and Price P-values are all zero or essentially zero, we have a strong case to reject the null hypothesis. However, the p-value for Urban is high, so we cannot reject the null hypothesis.

#### PART E : 
On the basis of your response to the previous question, fit a smaller model that only uses the predictors for which there is evidence of association with the outcome.

In [3]:
# Using R-style formulas to create another regression model without Urban
mod_2 = smf.ols(formula = "Sales ~ Price + US", data = df)
res_2 = mod_2.fit()
print(res_2.summary())

                            OLS Regression Results                            
Dep. Variable:                  Sales   R-squared:                       0.239
Model:                            OLS   Adj. R-squared:                  0.235
Method:                 Least Squares   F-statistic:                     62.43
Date:                Mon, 19 Sep 2022   Prob (F-statistic):           2.66e-24
Time:                        18:10:59   Log-Likelihood:                -927.66
No. Observations:                 400   AIC:                             1861.
Df Residuals:                     397   BIC:                             1873.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     13.0308      0.631     20.652      0.0

#### PART F : 
How well do the models in (a) and (e) fit the data?

If we look at the R-Squared values provided for each model, which represents the variation in the dependent variable that is predictable from the independent variable, we can see that both are around .24. That is a pretty low value, so both models would be considered average if not less than average.

#### PART G : 
Using the model from (e), obtain 95 % confidence intervals for the coefficient(s).

In [4]:
res_2.conf_int(alpha=0.05, cols=None)

0          1
Intercept  11.79032  14.271265
US[T.Yes]   0.69152   1.707766
Price      -0.06476  -0.044195

#### PART H : 
Design a model involving at least one interaction feature (e.g. X1*X2) and one transformed feature (e.g. X1^2.) Write your model mathematically and use simulation data to comment your findings on the interaction feature and transformed feature used in the model.  

*Ŷ = Intercept + ((CompPrice)X1 * (Price)X2) + (Income)X3^2*

In [5]:
# Using R-style formulas to create another regression model without Urban
mod_2 = smf.ols(formula = "Sales ~ CompPrice * Price + Income**2", data = df)
res_2 = mod_2.fit()
print(res_2.summary())

                            OLS Regression Results                            
Dep. Variable:                  Sales   R-squared:                       0.381
Model:                            OLS   Adj. R-squared:                  0.374
Method:                 Least Squares   F-statistic:                     60.66
Date:                Mon, 19 Sep 2022   Prob (F-statistic):           6.41e-40
Time:                        18:10:59   Log-Likelihood:                -886.58
No. Observations:                 400   AIC:                             1783.
Df Residuals:                     395   BIC:                             1803.
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
                      coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------
Intercept           4.9309      3.508     

In the above finding, we can see the interactive feature, which are the price and the competition price, the coefficient -1.382^-6, therefore there is very little change in price and comprice based around change in units. The p-value is .995, which makes the variables of price and competition price highly correlated. </br>
Alternative, we can see the transformative feature, which is income squared, the coefficient is .0153, therefore there is still very little change in income ased around change in units. The p-value is close to zero, therefore there is no correlation between income and units.